pkg/streamingpromql/operators/functions/quantile.go

```diff
@@ -0,0 +1,322 @@
+// explicitly):
+//
+// If the histogram has 0 observations, NaN is returned.
+//
+// If q<0, -Inf is returned.
+//
+// If q>1, +Inf is returned.
+//
+// If q is NaN, NaN is returned.
+func histogramQuantile(q float64, h *histogram.FloatHistogram) float64 {
+	if q < 0 {
+		return math.Inf(-1)
+	}
+	if q > 1 {
+		return math.Inf(+1)
+	}
+
+	if h.Count == 0 || math.IsNaN(q) {
+		return math.NaN()
+	}
+
+	var (
+		bucket histogram.Bucket[float64]
+		count  float64
+		it     histogram.BucketIterator[float64]
+		rank   float64
+	)
+
+	// if there are NaN observations in the histogram (h.Sum is NaN), use the forward iterator
+	// if the q < 0.5, use the forward iterator
+	// if the q >= 0.5, use the reverse iterator
+	if math.IsNaN(h.Sum) || q < 0.5 {
+		it = h.AllBucketIterator()
+		rank = q * h.Count
+	} else {
+		it = h.AllReverseBucketIterator()
+		rank = (1 - q) * h.Count
+	}
+
+	for it.Next() {
+		bucket = it.At()
+		if bucket.Count == 0 {
+			continue
+		}
+		count += bucket.Count
+		if count >= rank {
+			break
+		}
+	}
+	if !h.UsesCustomBuckets() && bucket.Lower < 0 && bucket.Upper > 0 {
+		switch {
+		case len(h.NegativeBuckets) == 0 && len(h.PositiveBuckets) > 0:
+			// The result is in the zero bucket and the histogram has only
+			// positive buckets. So we consider 0 to be the lower bound.
+			bucket.Lower = 0
+		case len(h.PositiveBuckets) == 0 && len(h.NegativeBuckets) > 0:
+			// The result is in the zero bucket and the histogram has only
+			// negative buckets. So we consider 0 to be the upper bound.
+			bucket.Upper = 0
+		}
+	} else if h.UsesCustomBuckets() {
+		if bucket.Lower == math.Inf(-1) {
+			// first bucket, with lower bound -Inf
+			if bucket.Upper <= 0 {
+				return bucket.Upper
+			}
+			bucket.Lower = 0
+		} else if bucket.Upper == math.Inf(1) {
+			// last bucket, with upper bound +Inf
+			return bucket.Lower
+		}
+	}
+	// Due to numerical inaccuracies, we could end up with a higher count
+	// than h.Count. Thus, make sure count is never higher than h.Count.
+	if count > h.Count {
+		count = h.Count
+	}
+	// We could have hit the highest bucket without even reaching the rank
+	// (this should only happen if the histogram contains observations of
+	// the value NaN), in which case we simply return the upper limit of the
+	// highest explicit bucket.
+	if count < rank {
+		return bucket.Upper
+	}
+
+	// NaN observations increase h.Count but not the total number of
+	// observations in the buckets. Therefore, we have to use the forward
+	// iterator to find percentiles. We recognize histograms containing NaN
+	// observations by checking if their h.Sum is NaN.
+	if math.IsNaN(h.Sum) || q < 0.5 {
+		rank -= count - bucket.Count
+	} else {
+		rank = count - rank
+	}
+
+	// TODO(codesome): Use a better estimation than linear.
+	return bucket.Lower + (bucket.Upper-bucket.Lower)*(rank/bucket.Count)
+}
+
+// coalesceBuckets merges buckets with the same upper bound.
```

Comment:
When does this happen in practice?

Reply:
I'm not sure that it would to be honest. However this file comes from Prometheus and I'd rather not make optimisations or changes in this PR. Might be something to come back to in the future if we are going to maintain our own quantile.

Reply:
I'm going to mark this as resolved. If you disagree, feel free to re-open :-).

Reply:
Perhaps we could contribute these improvements upstream? I also suspect that given we've copied these files, they will naturally diverge over time (unless we enforce that they remain in sync somehow), so I'm OK with making changes to them.